test/lib/DogpileSpec.scala

```diff
@@ -0,0 +1,39 @@
+package lib
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.{AtomicLong, LongAdder}
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+class DogpileSpec extends AnyFlatSpec with Matchers {
+  it should "not concurrently execute the side-effecting function" in {
```

Comment:
I added this test while trying to debug this issue. It never reproduced the problem - because the problem _wasn't_ in `Dogpile` - and due to the unpredictable nature of concurrency problems, the test wouldn't have been guaranteed to spot one if there _was_ one - but I guess it's a reasonable statement of intent.